adapters/rest-adapter-generator/src/main/java/uk/gov/justice/services/adapters/rest/generator/MediaTypeMapperGenerator.java

```diff
@@ -0,0 +1,102 @@
+
+    @Override
+    TypeSpec generateFor(final Resource resource) {
+
+        final String className = mapperClassNameOf(resource);
+        return classBuilder(className)
+                .addModifiers(PUBLIC)
+                .superclass(ClassName.get(BasicActionMapper.class))
+                .addAnnotation(builder(Named.class)
+                        .addMember("value", "$S", className).build())
+                .addMethod(constructorBuilder()
+                        .addModifiers(PUBLIC)
+                        .addCode(mapperConstructorCodeFor(resource))
+                        .build())
+                .build();
+    }
+
+    private CodeBlock mapperConstructorCodeFor(final Resource resource) {
+        final CodeBlock.Builder constructorCode = CodeBlock.builder();
+
+        resource.getActions().values().forEach(ramlAction -> {
+            constructorCode.add(statementsOf(ramlAction));
+        });
+
+        return constructorCode.build();
+    }
+
+    /**
+     * Process the body or bodies for each httpAction.
+     *
+     * @param ramlAction the httpAction to statementsOf
+     * @return the list of {@link CodeBlock} that represents each method for the httpAction
+     */
+    private CodeBlock statementsOf(final Action ramlAction) {
+        if (!ramlAction.hasBody()) {
+            return processActionBody(ramlAction, responseMimeTypesOf(ramlAction), this::resourceMethodNameSingle);
+        } else {
+            return processActionBody(ramlAction, requestMimeTypes(ramlAction), this::resourceMethodNameMultiple);
+        }
+    }
+
+    private CodeBlock processActionBody(final Action ramlAction, final Collection<MimeType> mimeTypes,
+                                        final BiFunction<Action, MimeType, String> resourceMethodName) {
+        final CodeBlock.Builder codeBlock = CodeBlock.builder();
+
+        mimeTypes.stream()
+                .sorted(byMimeTypeOrder())
+                .forEach(mimeType ->
+                        codeBlock.addStatement("add($S, $S, $S)",
+                                resourceMethodName.apply(ramlAction, mimeType),
+                                mimeType.getType(),
+                                nameFrom(mimeType.getType()))
+                );
+
+        return codeBlock.build();
+    }
+
+    private String resourceMethodNameMultiple(final Action ramlAction, final MimeType mimeType) {
+        return buildResourceMethodName(ramlAction, mimeType);
+    }
+
+    private String resourceMethodNameSingle(final Action ramlAction, final MimeType mimeType) {
+        return buildResourceMethodNameWithNoMimeType(ramlAction);
+    }
+
+    private Collection<MimeType> requestMimeTypes(final Action ramlAction) {
+        return ramlAction.getBody().values();
+    }
+
+    private String mapperClassNameOf(final Resource resource) {
+        return new StringBuilder().append(resourceImplementationNameOf(resource)).append("ActionMapper").toString();
```

Comment:
This is specific to the mapper class name and repeated, so I have moved it inside Names utils class.